### PR TITLE
Autoscaling capability to opensearch-dashboards helm chart

### PR DESCRIPTION
**Pull request by prudhvigodithi**
Signed-off-by: pgodithi <pgodithi@amazon.com>

Description
This PR adds autoscaling capability to opensearch-dashboards helm chart, this autoscaling field exists but was redundant as there was no necessary configs under templates.

Issues Resolved
https://github.com/opensearch-project/helm-charts/issues/230

Check List

- [x]  Commits are signed per the DCO using --signoff

For any changes to files within Helm chart directories:

- [x]  Helm chart version bumped

 

- [x] Helm chart CHANGELOG.md updated to reflect change

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/helm-charts/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by prudhvigodithi** [comment]
Hello @TheAlgo and @DandyDeveloper can you review this PR when you get a chance?

File: charts/opensearch-dashboards/CHANGELOG.md [hunk 1]
```diff
@@ -146,6 +159,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Security
 
 [Unreleased]: https://github.com/opensearch-project/helm-charts/compare/opensearch-dashboards-1.3.0...HEAD
```
Comment (peterzhuamazon) [inline charts/opensearch-dashboards/CHANGELOG.md:161]: This should be 1.3.1 to head.

File: charts/opensearch-dashboards/README.md [hunk 2]
```diff
@@ -74,3 +74,4 @@
...
 | `updateStrategy`                   | The [updateStrategy][] for the StatefulSet. By default Kubernetes will wait for the cluster to be green after upgrading each pod. Setting this to `OnDelete` will allow you to manually delete each pod during upgrades                                   | `RollingUpdate`                                 |
 | `extraObjects`                     | Array of extra K8s manifests to deploy                                                                                                                                                                                                                    | list `[]`                                       |
+| `autoscaling`                          | Prerequisite: Install/Configure metrics server. Configurable pod autoscaling stratergy to scale based on `targetCPUUtilizationPercentage`, configure `minReplicas` and `maxReplicas` for desired scaling                                                                                                                                                        | false                             |
```
Comment (TheAlgo) [inline charts/opensearch-dashboards/README.md:77]: Good to attach the link in the README as well for the installation

File: charts/opensearch-dashboards/values.yaml [hunk 3]
```diff
@@ -146,11 +146,12 @@ resources:
     memory: "512M"
 
 autoscaling:
-  enabled: false
+  # This requires metrics server to be installed, to install use kubectl apply -f https://github.com/kubernetes-sigs/metrics-server/releases/latest/download/components.yaml
+  # See https://github.com/kubernetes-sigs/metrics-server
+  enabled: true
```
Comment (TheAlgo) [inline charts/opensearch-dashboards/values.yaml:151]: IMO Lets not keep it enabled by default. For resources we can keep it silenced, this should not blast someone's cluster. Also this will fail things if metric server is not installed in the setup.

File: charts/opensearch-dashboards/CHANGELOG.md [hunk 4]
```diff
@@ -13,6 +13,19 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed
 ### Security
 ---
+## [1.3.1]
+### Added
+- Add pod autoscaling capability.
+- Added autoscaling configuration parameters in readme.
+- Added autoscaling.yaml under templates.
+### Changed
+- Updated the chart version to 1.2.3
```
Comment (TheAlgo) [inline charts/opensearch-dashboards/CHANGELOG.md:22]: Typo: Chart version should go `1.3.1`

File: charts/opensearch-dashboards/CHANGELOG.md [hunk 5]
```diff
@@ -13,6 +13,19 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed
 ### Security
 ---
+## [1.3.1]
+### Added
+- Add pod autoscaling capability.
+- Added autoscaling configuration parameters in readme.
```
Comment (TheAlgo) [inline charts/opensearch-dashboards/CHANGELOG.md:19]: Nit: Can we maintain a common tense here for better readability?

File: charts/opensearch-dashboards/values.yaml [hunk 6]
```diff
@@ -146,11 +146,12 @@ resources:
...
   minReplicas: 1
-  maxReplicas: 100
+  maxReplicas: 10
   targetCPUUtilizationPercentage: 80
-  # targetMemoryUtilizationPercentage: 80
```
Comment (TheAlgo) [inline charts/opensearch-dashboards/values.yaml:153]: Can we add memory as well apart from CPU here because this is something I have seen clusters experiencing. Though I am not really sure.

@peterzhuamazon @DandyDeveloper What's your opinion on this?

**Review by TheAlgo — CHANGES_REQUESTED** [review]

File: charts/opensearch-dashboards/values.yaml (hunk 6 above)
Comment (prudhvigodithi) [reply charts/opensearch-dashboards/values.yaml:153]: Hey @TheAlgo, @DandyDeveloper, @peterzhuamazon 
I have added memory as well, I have seen cluster issues with pods getting terminated OOM killed error, so I have pushed hpa with api Autoscaling/v1, this is stable version that has only CPU based scaling, the other supported API is Autoscaling/v2beta2 this has both CPU and memory but in beta version, but from doc I see its now in stable release https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/ (v2beta1 is deprecated), so we should be good to use this v2beta2 API, I have added targetMemoryUtilizationPercentage in my latest commit.
Screen Shot 2022-03-20 at 7 09 32 AM

File: charts/opensearch-dashboards/CHANGELOG.md (hunk 5 above)
Comment (prudhvigodithi) [reply charts/opensearch-dashboards/CHANGELOG.md:19]: Fixed in latest commit.

File: charts/opensearch-dashboards/CHANGELOG.md (hunk 4 above)
Comment (prudhvigodithi) [reply charts/opensearch-dashboards/CHANGELOG.md]: Fixed in latest commit.

File: charts/opensearch-dashboards/values.yaml (hunk 3 above)
Comment (prudhvigodithi) [reply charts/opensearch-dashboards/values.yaml]: Fixed in latest commit.

File: charts/opensearch-dashboards/README.md (hunk 2 above)
Comment (prudhvigodithi) [reply charts/opensearch-dashboards/README.md]: Fixed in latest commit.

File: charts/opensearch-dashboards/CHANGELOG.md (hunk 1 above)
Comment (prudhvigodithi) [reply charts/opensearch-dashboards/CHANGELOG.md]: Fixed in latest commit.

File: charts/opensearch-dashboards/values.yaml (hunk 6 above)
Comment (DandyDeveloper) [reply charts/opensearch-dashboards/values.yaml:153]: @prudhvigodithi @TheAlgo For opensearch-dashboards I don't see this as an issue. 

What's the feature state and planned implementation for `v1/v2beta2`? I can't find any reference on when it was introduced.  https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#horizontalpodautoscaler-v2beta2-autoscaling

We might want to be careful as some people using old clusters won't be able to use the HPA if it's only available in 1.22+ (But I can't find exactly when this was introduced).

Can we triple check that? Otherwise I'd keep CPU targets for now and use `v1`

Comment (prudhvigodithi) [reply charts/opensearch-dashboards/values.yaml:153]: Hey @DandyDeveloper, good point, I can see this v2beta2 was introduced in k8s 1.12 release notes
https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.12.md

The cluster which I have tested is v1.21.5 that has all autoscaling/v1, autoscaling/v2beta1, autoscaling/v2beta2
```
kubectl api-versions | grep autoscaling
autoscaling/v1
autoscaling/v2beta1
autoscaling/v2beta2
```

v2beta1 is deprecated in 1.19 and you should use v2beta2 going forward. https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.19.md

Also on Default k8s docs I can see this 
https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/
FEATURE STATE: Kubernetes v1.23 [stable]
(the autoscaling/v2beta2 API version previously provided this ability as a beta feature)

But looks like GKE has some issues and resolved in 1.15.5
https://issuetracker.google.com/issues/135624588?pli=1

So to summarize autoscaling/v2beta2 should be available from k8s version 1.12, GKE 1.15.5.
I'm ok to go with v1 and keep CPU alone to scale excluding memory for now, assuming there are chances some users can be before 1.12 version, or I can go with autoscaling/v2beta2 adding a note saying that it should be above 1.12 for this to work @TheAlgo and @DandyDeveloper

**Comment by peterzhuamazon** [comment]
@DandyDeveloper @TheAlgo please review again and see if @prudhvigodithi has addressed all your concerns.
Thanks.

File: charts/opensearch-dashboards/values.yaml (hunk 6 above)
Comment (TheAlgo) [reply charts/opensearch-dashboards/values.yaml:153]: @prudhvigodithi In that case lets keep CPU for now. Lets not add that, it might create unnecessary confusion for the users.

Comment (peterzhuamazon) [reply charts/opensearch-dashboards/values.yaml:153]: @prudhvigodithi ^^

Comment (prudhvigodithi) [reply charts/opensearch-dashboards/values.yaml:153]: Hey @peterzhuamazon and @TheAlgo I have reverted to v1, let's merge this and open an issue to add even memory as a scaler metric, and we can add updates to that issue about v2 autoscaling version. 
Thank you

File: charts/opensearch-dashboards/CHANGELOG.md [hunk 7]
```diff
@@ -13,6 +13,20 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+- Added autoscaling.yaml under templates.
+### Changed
+- Updated the chart version to 1.3.2
+- Changed values.yaml to reflect autoscaling changes.
+### Deprecated
+### Removed
+### Fixed
+### Security
+---
+=======
```
Comment (peterzhuamazon) [inline charts/opensearch-dashboards/CHANGELOG.md:29]: What is this?

Comment (peterzhuamazon) [reply charts/opensearch-dashboards/CHANGELOG.md:29]: I think you forget to remove it during resolving conflicts.

File: charts/opensearch-dashboards/CHANGELOG.md [hunk 8]
```diff
@@ -154,7 +168,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed
 ### Security
 
-[Unreleased]: https://github.com/opensearch-project/helm-charts/compare/opensearch-dashboards-1.3.1...HEAD
+[Unreleased]: https://github.com/opensearch-project/helm-charts/compare/opensearch-dashboards-1.3.2...HEAD
+[1.3.2]: https://github.com/opensearch-project/helm-charts/compare/opensearch-1.3.0...opensearch-1.3.2
```
Comment (peterzhuamazon) [inline charts/opensearch-dashboards/CHANGELOG.md:172]: 1.3.1 not 1.3.0

File: charts/opensearch-dashboards/Chart.yaml [hunk 9]
```diff
@@ -15,7 +15,8 @@ type: application
 # This is the chart version. This version number should be incremented each time you make changes
 # to the chart and its templates, including the app version.
 # Versions are expected to follow Semantic Versioning (https://semver.org/)
-version: 1.3.1
+
```
Comment (peterzhuamazon) [inline charts/opensearch-dashboards/Chart.yaml:18]: Remove extra space that is added during resolving conflicts.

**Review by peterzhuamazon — APPROVED** [review]

**Comment by prudhvigodithi** [comment]
Hey @TheAlgo and @DandyDeveloper , can you please check this PR?
Thank you

File: charts/opensearch-dashboards/values.yaml (hunk 6 above)
Comment (prudhvigodithi) [reply charts/opensearch-dashboards/values.yaml:152]: Hey @alborotogarcia and @DandyDeveloper, can you please check this change that I have pushed regarding v1.
Thank you

**Review by TheAlgo — APPROVED** [review]
LGTM!
